### PR TITLE
initial project setup

### LINKED ISSUE #1: Create react app with TS
Use create-react-app with Typescript support

### PR DESCRIPTION
**Pull request by AhmetTanakol**
- Create a project scaffolding
- Configure Travis to run tests and lint check
- Setup Jest and tslint

Fixes #1 #2 #3 #4 #5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,2 @@
 # jsonforms-ui-schema-editor
+[![Coverage Status](https://coveralls.io/repos/github/AhmetTanakol/jsonforms-ui-schema-editor/badge.svg?branch=master)](https://coveralls.io/github/AhmetTanakol/jsonforms-ui-schema-editor?branch=master)
```
Comment (edgarmueller) [inline README.md]: This should point to upstream rather than your fork. ('ve enabled coveralls for upstream).

File: public/index.html [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    <!--
+      Notice the use of %PUBLIC_URL% in the tags above.
+      It will be replaced with the URL of the `public` folder during the build.
+      Only files inside the `public` folder can be referenced from the HTML.
+
+      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
+    <title>React App</title>
```
Comment (edgarmueller) [inline public/index.html]: Change title

File: public/manifest.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+  "short_name": "React App",
```
Comment (edgarmueller) [inline public/manifest.json]: Update short_name, name

File: tsconfig.test.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "module": "commonjs"
+  }
+}
```
Comment (edgarmueller) [inline tsconfig.test.json]: newline

File: src/ui-metaschema.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
+export const uiMetaSchema = {
```
Comment (edgarmueller) [inline src/ui-metaschema.ts]: Add $schema keyword for version 7

File: src/ui-metaschema.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+          { '$ref': '#/definitions/group' }
+        ]
+      }
+    },
+    'control': {
+      'type': 'object',
+      'properties': {
+        'type': {
+          'type': 'string',
+          'enum': ['Control']
```
Comment (edgarmueller) [inline src/ui-metaschema.ts]: `enum`s with just a single value should be replaced with `const`

File: src/ui-metaschema.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        },
+        'label': {
+          '$ref': '#/definitions/label'
+        },
+        'scope': {
+          '$ref': '#/definitions/scope',
+        },
+        'options': {
+          'type': 'object',
```
Comment (edgarmueller) [inline src/ui-metaschema.ts]: `options` should itself be a definition and it should allow for objects where values can have an arbitrary type

File: src/ui-metaschema.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+          'properties': {
+            'additionalProperties': {
+              'type': 'string'
+            }
+          }
+        },
+        'suggestion': {
```
Comment (edgarmueller) [inline src/ui-metaschema.ts]: Remove this, this would be an option

File: src/ui-metaschema.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+          'required': ['type', 'scope', 'expectedValue']
+        }
+      },
+      'required': ['effect', 'condition']
+    },
+    'scope': {
+      'type': 'string',
+      'pattern': '^#\\/properties\\/{1}'
+    },
+    'label-object': {
```
Comment (edgarmueller) [inline src/ui-metaschema.ts]: Replace hyphen with camelCase

File: src/validation.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import * as AJV from 'ajv';
+import { ErrorObject } from 'ajv';
+const ajv = new AJV({allErrors: true, verbose: true});
+
+const extractErrors = (errors: ErrorObject[]): any[] =>
```
Comment (edgarmueller) [inline src/validation.ts]: Remove references to `any` type, introduce respective interfaces if necessary,

File: test/validation.test.ts [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+
+import { validate } from '../src/validation';
+import {
+  schemaRule,
+  uischemaCategorization,
+  uischemaControl,
+  uischemaGroup,
+  uischemaHorizontalLayout,
+  uischemaVerticalLayout
+} from './exampleUISchemas/validSchemas';
```
Comment (edgarmueller) [inline test/validation.test.ts]: Use star import : `* as valid from ...`
References to actual schemas then can be prefixed with `valid` which is more explicit.

File: test/validation.test.ts [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+        },
+        scope: '#/properties/occupation'
+      }
+    ]
+  };
+
+  expect(validator(uischema)).toEqual([]);
+});
+
+test('invalid control element, invalid label', () => {
```
Comment (edgarmueller) [inline test/validation.test.ts]: Rename to 'layout label must be string'

File: test/validation.test.ts [hunk 13]
```diff
@@ -0,0 +1,586 @@
+/* tslint:disable:max-file-line-count */
```
Comment (edgarmueller) [inline test/validation.test.ts]: Rename file to uischemaValidation or simliar

File: test/validation.test.ts [hunk 14]
```diff
@@ -0,0 +1,586 @@
...
+          show: true,
+          text: 'Occupation',
+          hide: false
+        },
+        scope: '#/properties/occupation'
+      }
+    ]
+  };
+  const errors = validator(uischema);
+  expect(errors[0].message).toEqual('should be string');
```
Comment (edgarmueller) [inline test/validation.test.ts]: I'd rather use `toHaveLength` or a similar matcher; It's not immediately obvious why the message is `should be string`.

File: test/validation.test.ts [hunk 15]
```diff
@@ -0,0 +1,586 @@
...
+});
+
+test('invalid control element, label with additional properties', () => {
```
Comment (edgarmueller) [inline test/validation.test.ts]: Rename to 'label object does not allow additional properties'

File: test/validation.test.ts [hunk 16]
```diff
@@ -0,0 +1,586 @@
...
+  const uischema = {
+    elements: [
+      {
+        type: 'Control',
+        scope: '#/properties/name'
+      }
+    ]
+  };
+  const errors = validator(uischema);
+  expect(errors[0].message).toEqual('should have required property \'type\'');
```
Comment (edgarmueller) [inline test/validation.test.ts]: Use template strings to avoid escaping

File: test/validation.test.ts [hunk 17]
```diff
@@ -0,0 +1,586 @@
...
+        scope: '#/properties/occupation',
+        options: 'multi'
+      }
+    ]
+  };
+  const errors = validator(uischema);
+  expect(errors[0].message).toEqual('should be object');
+});
+
+test('invalid control element, invalid suggestion type', () => {
```
Comment (edgarmueller) [inline test/validation.test.ts]: Update this and move suggestion to `options` block

**Review by edgarmueller — CHANGES_REQUESTED** [review]
Looks good, I'd propose to fix the comments made regarding the UI schema, everything else also can be fixed later, given we create the respective issues.

**Comment by edgarmueller** [comment]
Also add a test case with a scope of nested properties

File: src/validation.ts (hunk 10 above)
Comment (AhmetTanakol) [reply src/validation.ts]: Do I need to create an interface for ui metaschema?

**Comment by AhmetTanakol** [comment]
updated

File: src/models/jsonSchema.ts [hunk 18]
```diff
@@ -0,0 +1,129 @@
+export interface JsonSchema {
```
Comment (edgarmueller) [inline src/models/jsonSchema.ts:1]: This interface mimics v4, but we are using v7

**Review by edgarmueller — CHANGES_REQUESTED** [review]
Update schema to v7

**Comment by AhmetTanakol** [comment]
I tried to migrate json schema from v04 to v07. I think we should look at properties `$schema` and `default` in jsonforms. Is `$schema` should be string or JsonSchema? (maybe both?). I don't know why but we use single quotation marks for type definition of `default` property in Jsonforms.

screen shot 2018-04-13 at 6 10 03 pm

**Review by edgarmueller — APPROVED** [review]

**Comment by edgarmueller** [comment]
Yep, good points, we'll have a look, otherwise LGTM

**Pull request merged**